modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersValidator.java

```diff
@@ -0,0 +1,168 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.http.netty4;
+
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.DefaultHttpRequest;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMessage;
+import io.netty.handler.codec.http.HttpRequest;
+
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.TriConsumer;
+import org.elasticsearch.http.HttpHeadersValidationException;
+import org.elasticsearch.http.HttpPreRequest;
+import org.elasticsearch.http.netty4.HttpHeadersValidator.ValidatableHttpHeaders.ValidationResult;
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.getHttpHeadersAsMap;
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.translateRequestMethod;
+
+/**
+ * Provides utilities used to hook into the netty pipeline and run a validation check on each HTTP request's headers.
+ * See also {@link Netty4HttpHeaderValidator}.
+ */
+public final class HttpHeadersValidator {
+
+    /**
+     * An async HTTP headers validator function that receives as arguments part of the incoming HTTP request
+     * (except the body contents, see {@link HttpPreRequest}), as well as the netty channel that the request is
+     * being received over, and must then call the {@code ActionListener#onResponse} method on the listener parameter
+     * in case the validation is to be considered successful, or otherwise call {@code ActionListener#onFailure}.
+     */
+    private final TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResult>> validator;
+
+    public HttpHeadersValidator(TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResult>> validator) {
+        this.validator = validator;
+    }
+
+    /**
+     * Supplies a netty {@code ChannelInboundHandler} that runs the {@link #validator} on the HTTP request headers.
+     * The HTTP headers of the to be validated {@link HttpRequest} must be wrapped by the special {@link ValidatableHttpHeaders},
+     * see {@link #wrapAsValidatableMessage(HttpMessage)}.
+     */
+    public Netty4HttpHeaderValidator getValidatorInboundHandler() {
+        return new Netty4HttpHeaderValidator((httpRequest, channel, listener) -> {
+            if (httpRequest.headers() instanceof ValidatableHttpHeaders validatableHttpHeaders) {
+                // make sure validation only runs on properly wrapped "validatable" headers implementation
+                validator.apply(asHttpPreRequest(httpRequest), channel, ActionListener.wrap(validationResult -> {
+                    validatableHttpHeaders.markAsSuccessfullyValidated(validationResult);
```

Comment:
ideally calling into authentication and handling the result would happen inside `Netty4HttpHeaderValidator`  (and be called called authentication.

Reply:
I think between [e7e15fd](https://github.com/elastic/elasticsearch/pull/95112/commits/e7e15fdc989e278f6068fc2ea791448acab83d27) and [625e56c](https://github.com/elastic/elasticsearch/pull/95112/commits/625e56c38b302bd5f7e788ad18f0e183e5c40f1a) this achieves the point you've raised here.